src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,190 @@
+                        killProcessOnCancel: true,
+                        cancellationToken: cancellationToken);
+
+                executionContext.Output(StringUtil.Loc("RobocopyBasedPublishArtifactTaskExitCode", exitCode));
+
+                if (exitCode >= 8)
+                {
+                    throw new Exception(StringUtil.Loc("RobocopyBasedPublishArtifactTaskFailed", exitCode));
+                }
+            }
+        }
+
+        private async Task DownloadFileShareArtifactAsync(string sourcePath, string destPath, int parallelCount, CancellationToken cancellationToken, IEnumerable<Func<string, bool>> minimatchFuncs = null)
+        {
+            var trimChars = new[] { '\\', '/' };
+
+            // If user has specified a relative folder in the drop, change the drop location itself. 
+            sourcePath = Path.Combine(sourcePath.TrimEnd(trimChars));
+
+            IEnumerable<string> files =
+                new DirectoryInfo(sourcePath).EnumerateFiles("*", SearchOption.AllDirectories)
+                    .Select(path => path.FullName);
+
+            var parallelism = new ExecutionDataflowBlockOptions()
+            {
+                MaxDegreeOfParallelism = parallelCount,
+                BoundedCapacity = 2 * parallelCount,
+                CancellationToken = cancellationToken
+            };
+
+            var actionBlock = NonSwallowingActionBlock.Create<string>(
+                action: async file =>
+                {
+                    if (minimatchFuncs == null || minimatchFuncs.Any(match => match(file))) 
+                    {
+                        string tempPath = Path.Combine(destPath, Path.GetRelativePath(sourcePath, file));
+                        this.context.Output(StringUtil.Loc("CopyFileToDestination", file, tempPath));
+                        FileInfo tempFile = new System.IO.FileInfo(tempPath);
+                        using (StreamReader fileReader = GetFileReader(file))
+                        {
+                            await WriteStreamToFile(
+                                fileReader.BaseStream,
+                                tempFile.FullName,
+                                DefaultStreamBufferSize,
+                                cancellationToken);
+                        }
+                    }
+                },
+                dataflowBlockOptions: parallelism);
+                
+                await actionBlock.SendAllAndCompleteAsync(files, actionBlock, cancellationToken);
+        }
+        
+        private async Task WriteStreamToFile(Stream stream, string filePath, int bufferSize, CancellationToken cancellationToken)
+        {
+            ArgUtil.NotNull(stream, nameof(stream));
+            ArgUtil.NotNullOrEmpty(filePath, nameof(filePath));
+
+            EnsureDirectoryExists(Path.GetDirectoryName(filePath));
+            using (var targetStream = new FileStream(filePath, FileMode.Create, FileAccess.Write, FileShare.None, bufferSize, true))
```

Comment:
explicitly provide the parameter name for reviewers :)

Reply:
outdated